File: 2jin8/공통/BOJ_2096.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class BOJ_2096 {

	static int N;
	static int[][] arr, minDP, maxDP;

	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		N = Integer.parseInt(br.readLine());
		arr = new int[N + 1][3];
		minDP = new int[N + 1][3];
		maxDP = new int[N + 1][3];
		for (int i = 1; i <= N; i++) {
			StringTokenizer st = new StringTokenizer(br.readLine());
			for (int j = 0; j < 3; j++) {
				arr[i][j] = Integer.parseInt(st.nextToken());
			}
		}

		for (int i = 1; i <= N; i++) {
			minDP[i][0] = Math.min(minDP[i - 1][0], minDP[i - 1][1]) + arr[i][0];
			minDP[i][1] = Math.min(minDP[i - 1][0], Math.min(minDP[i - 1][1], minDP[i - 1][2])) + arr[i][1];
			minDP[i][2] = Math.min(minDP[i - 1][1], minDP[i - 1][2]) + arr[i][2];

			maxDP[i][0] = Math.max(maxDP[i - 1][0], maxDP[i - 1][1]) + arr[i][0];
			maxDP[i][1] = Math.max(maxDP[i - 1][0], Math.max(maxDP[i - 1][1], maxDP[i - 1][2])) + arr[i][1];
			maxDP[i][2] = Math.max(maxDP[i - 1][1], maxDP[i - 1][2]) + arr[i][2];
		}

		int min = Integer.MAX_VALUE, max = Integer.MIN_VALUE;
		for (int i = 0; i < 3; i++) {
			min = Math.min(min, minDP[N][i]);
			max = Math.max(max, maxDP[N][i]);
		}
		System.out.println(max + " " + min);
	}
}

File: 2024년/24년 10월/10월 2주차 완료/hongjun/03/BOJ_2096.py
N = int(input())

max_dp = [0] * 3
min_dp = [0] * 3

tmp1 = [0] * 3
tmp2 = [0] * 3

for i in range(N):
    a, b, c = map(int, input().split())
    # 최대값
    for j in range(3):
        if j == 0:
            tmp1[j] = max(a + max_dp[j], a + max_dp[j+1])

        elif j == 1:
            tmp1[j] = max(b + max_dp[j-1], b + max_dp[j], b + max_dp[j+1])

        elif j == 2:
            tmp1[j] = max(c + max_dp[j - 1], c + max_dp[j])

    # 최소값
    for j in range(3):
        if j == 0:
            tmp2[j] = min(a + min_dp[j], a + min_dp[j + 1])

        elif j == 1:
            tmp2[j] = min(b + min_dp[j - 1], b + min_dp[j], b + min_dp[j + 1])

        elif j == 2:
            tmp2[j] = min(c + min_dp[j - 1], c + min_dp[j])


    # dp 갱신
    for k in range(3):
        max_dp[k] = tmp1[k]
        min_dp[k] = tmp2[k]

print(max(max_dp), min(min_dp))